eclair-core/src/main/scala/fr/acinq/eclair/Features.scala

```diff
@@ -19,63 +19,115 @@ package fr.acinq.eclair
 
-  val CHANNEL_RANGE_QUERIES_EX_BIT_MANDATORY = 10
-  val CHANNEL_RANGE_QUERIES_EX_BIT_OPTIONAL = 11
+  case object VariableLengthOnion extends Feature {
+    val rfcName = "var_onion_optin"
+    val mandatory = 8
+  }
 
-  val PAYMENT_SECRET_MANDATORY = 14
-  val PAYMENT_SECRET_OPTIONAL = 15
+  case object ChannelRangeQueriesExtended extends Feature {
+    val rfcName = "gossip_queries_ex"
+    val mandatory = 10
+  }
 
-  val BASIC_MULTI_PART_PAYMENT_MANDATORY = 16
-  val BASIC_MULTI_PART_PAYMENT_OPTIONAL = 17
+  case object PaymentSecret extends Feature {
+    val rfcName = "payment_secret"
+    val mandatory = 14
+  }
+
+  case object BasicMultiPartPayment extends Feature {
+    val rfcName = "basic_mpp"
+    val mandatory = 16
+  }
 
   // TODO: @t-bast: update feature bits once spec-ed (currently reserved here: https://github.com/lightningnetwork/lightning-rfc/issues/605)
-  // We're not advertizing these bits yet in our announcements, clients have to assume support.
+  // We're not advertising these bits yet in our announcements, clients have to assume support.
   // This is why we haven't added them yet to `areSupported`.
-  val TRAMPOLINE_PAYMENT_MANDATORY = 50
-  val TRAMPOLINE_PAYMENT_OPTIONAL = 51
+  case object TrampolinePayment extends Feature {
+    val rfcName = "trampoline_payment"
+    val mandatory = 50
+  }
+
+  // Features may depend on other features, as specified in Bolt 9.
+  private val featuresDependency = Map(
+    ChannelRangeQueriesExtended -> (ChannelRangeQueries :: Nil),
+    PaymentSecret -> (VariableLengthOnion :: Nil),
+    BasicMultiPartPayment -> (PaymentSecret :: Nil),
+    TrampolinePayment -> (PaymentSecret :: Nil)
+  )
+
+  case class FeatureException(message: String) extends IllegalArgumentException(message)
+
+  def validateFeatureGraph(features: BitVector): Option[FeatureException] = featuresDependency.collectFirst {
+    case (feature, dependencies) if hasFeature(features, feature, None) && dependencies.exists(d => !hasFeature(features, d, None)) =>
+      FeatureException(s"${features.toBin} sets $feature but is missing a dependency (${dependencies.filter(d => !hasFeature(features, d, None)).mkString(" and ")})")
+  }
+
+  def validateFeatureGraph(features: ByteVector): Option[FeatureException] = validateFeatureGraph(features.bits)
 
   // Note that BitVector indexes from left to right whereas the specification indexes from right to left.
   // This is why we have to reverse the bits to check if a feature is set.
 
-  def hasFeature(features: BitVector, bit: Int): Boolean = if (features.sizeLessThanOrEqual(bit)) false else features.reverse.get(bit)
+  private def hasFeature(features: BitVector, bit: Int): Boolean = if (features.sizeLessThanOrEqual(bit)) false else features.reverse.get(bit)
 
-  def hasFeature(features: ByteVector, bit: Int): Boolean = hasFeature(features.bits, bit)
+  def hasFeature(features: BitVector, feature: Feature, support: Option[FeatureSupport]): Boolean = support match {
```

Comment:
I think that's a rare case where I would put a default value for field `support`, because most of the time we just want to know if a feature is supported, without more details.

```suggestion
  def hasFeature(features: BitVector, feature: Feature, support: Option[FeatureSupport] = None): Boolean = support match {
```

Reply:
Maybe also `sed s/hasFeature(features, bit, None)/hasFeature(features, bit)/g` in the rest of the code?

Reply:
Done!